Add Dict concept and exercise

ceddlyburge opened the pull request:
Fixes #425

mpizenberg commented:
Hi, I haven't dug into this, but why mention `filter` in the comment above about removing elements and iterating? I thought `Dict.remove` and `Dict.map` was rather obvious for those, am I missing something?

ceddlyburge commented:
> Hi, I haven't dug into this, but why mention `filter` in the comment above about removing elements and iterating? I thought `Dict.remove` and `Dict.map` was rather obvious for those, am I missing something?

HI Matthieu, Dict.map and Dict.remove do make more sense in some ways, but it's difficult to make an engaging example. If you look at the other tracks then they basically just have functions called, say `remove` that you are supposed to fill in using `Dict.remove`, which seems uninteresting. 

Also Dict.remove is something that I rarely use, so I think it is ok to have it mentioned in the docs rather than having to actually use it in the exercise.

Dict.map would be useful, but again, I think it is better to have an interesting exercise rather than to shoe horn everything in. It doesn't make sense to ask the student to write every Dict function I think. It probably could be worked in though ...

mpizenberg commented:
Maybe we are understanding the "student should know how to remove an
element from a dict" differently. To me, the list of expected learning
objectives do not correspond 1 to 1 to actual tasks or functions to
implement. Instead they should be required in one of the tasks. This mean
designing on task such that removing an element is necessary. But this does
not mean we ask students to implement a function to remove elements from a
dict. Which as you've pointed out is rather meaningless if that's just
writing: remove = Dict.remove

On Wed, Mar 16, 2022, 12:48 Cedd Burge ***@***.***> wrote:

> Hi, I haven't dug into this, but why mention filter in the comment above
> about removing elements and iterating? I thought Dict.remove and Dict.map
> was rather obvious for those, am I missing something?
>
> HI Matthieu, Dict.map and Dict.remove do make more sense in some ways, but
> it's difficult to make an engaging example. If you look at the other tracks
> then they basically just have functions called, say remove that you are
> supposed to fill in using Dict.remove, which seems uninteresting.
>
> Also Dict.remove is something that I rarely use, so I think it is ok to
> have it mentioned in the docs rather than having to actually use it in the
> exercise.
>
> Dict.map would be useful, but again, I think it is better to have an
> interesting exercise rather than to shoe horn everything in. It doesn't
> make sense to ask the student to write every Dict function I think. It
> probably could be worked in though ...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/elm/pull/469#issuecomment-1069040410>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWFOCLJE3H27JVRDQ3JQSDVAHDALANCNFSM5QZYFKEQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

jiegillet commented on exercises/concept/top-scorers/.meta/Exemplar.elm:
```diff
@@ -0,0 +1,59 @@
+    --     updatePlayer existingPlayer =
+    --         case existingPlayer of
+    --             Just player ->
+    --                 { player | goals = player.goals + 1 }
```
The types don't match here, you are returning a record rather than an `Int`, and also `updatePlayer` should be of type `Maybe v -> Maybe v`.

ceddlyburge replied:
Yep, I haven't tried to run this code or anything, I'll fix that up later!

jiegillet commented on exercises/concept/top-scorers/.meta/Exemplar.elm:
```diff
@@ -0,0 +1,59 @@
+formatPlayer playerName = 
+    Debug.todo "implement formatPlayer function, for use on the player profile page"
+    -- Dict.get playerName
+    -- |> Maybe.map (\goalCount -> playerName ++ ": " + (toString goalCount)
```
`toString`? Is this Elm 0.18? haha

ceddlyburge replied:
Yep, I haven't tried to run this code or anything, I'll fix that up later!

jiegillet reviewed (commented):
Looks fine, it covers most sections of the `Dict` docs: build, query, lists, transform, but combine is missing. Maybe we could have football club merge together (`union`) or getting players who were present in two seasons (`intersect`) or something more convoluted (`merge`)?

ceddlyburge commented:
> Maybe we could have football club merge together (union) or getting players who were present in two seasons (intersect) or something more convoluted (merge)?

Good idea, I think it will be relatively easy to work in a combining task. I'll add it to the list of things to learn as well.

ceddlyburge commented:
Hi @mpizenberg  @jiegillet , I've finished the dict concept and exercise now, can you take a look and let me know what you think?

Thanks, Cedd

jiegillet commented on exercises/concept/top-scorers/.docs/introduction.md:
```diff
@@ -0,0 +1,55 @@
+
+A [`Dict`][dict] in Elm is an immutable dictionary of zero or more key-value pairs.
+
+Type annotations for dicts can be defined as follows
```
```suggestion
Type annotations for dicts are be written as follows
```

jiegillet commented on exercises/concept/top-scorers/.docs/introduction.md:
````diff
@@ -0,0 +1,55 @@
+Dict Int String --> a dict with Int keys and String values
+```
+
+Dicts can be defined as follows:
````
```suggestion
Dicts can be created as follows:
```

jiegillet commented on exercises/concept/top-scorers/.docs/introduction.md:
````diff
@@ -0,0 +1,55 @@
+twoValues = Dict.fromList [ ( 0, "Alice" ), ( 1, "Bob" ) ] --> Dict Int String
+```
+
+Items can be retrieved using `get`.
````
```suggestion
Items can be retrieved using `get`.
As a key may or may not be present in the dict, the result will be wrapped in a `Maybe`.
```

jiegillet commented on exercises/concept/top-scorers/.docs/introduction.md:
````diff
@@ -0,0 +1,55 @@
+keyNotPresent = Dict.get 1 --> Nothing
+```
+
+Items can be added using `insert`.
````
```suggestion
Items can be added using `insert`. 
Items for keys that are already in the dict get replaced.
```

jiegillet commented on exercises/concept/top-scorers/.docs/introduction.md:
````diff
@@ -0,0 +1,55 @@
+
+```elm
+alice = Dict.fromList [ ( 0, "Alice" ) ]
+aliceUpperCase = Dict.update 0 (Maybe.map toUpper)  --> ( 0, "ALICE" )
````
```suggestion
aliceUpperCase = Dict.update 0 (Maybe.map toUpper)  --> Dict.fromList [ ( 0, "ALICE" ) ]
aliceNoCase = Dict.update 0 (\_ -> Nothing)  --> Dict.empty
bobUpperCase = Dict.update 1 (\_ -> Just "BOB") --> Dict.fromList [ ( 0, "Alice" ), ( 1, "BOB" ) ]
bobNoCase = Dict.update 1 (\_ -> Nothing) --> Dict.fromList [ ( 0, "Alice" ) ]
```

jiegillet commented on exercises/concept/top-scorers/.docs/introduction.md:
````diff
@@ -0,0 +1,55 @@
+alice = Dict.fromList [ ( 0, "Alice" ) ]
+empty = Dict.remove 0 --> Dict.empty
+```
+
````
It's probably worth introducing `map`, `filter` and `merge` since we expect the students to use them.

jiegillet commented on exercises/concept/top-scorers/.meta/Exemplar.elm:
```diff
@@ -0,0 +1,61 @@
+module TopScorers exposing (..)
+
+import Dict as Dict exposing (Dict)
```
```suggestion
import Dict exposing (Dict)
```

jiegillet commented on exercises/concept/top-scorers/.meta/design.md:
```diff
@@ -0,0 +1,29 @@
+# Design
```
Would you like to add something for the analyzer?
It's a bit tricky because there are often several ways to solve a problem, but for the ones we are sure there is a best practice, we could enforce that (For example, making sure `filter` is used in `removeInsignificantPlayers` or `merge` in `combineGames`)

ceddlyburge replied:
Good idea.

How about this for a list of analyser checks?

- `filter` must be used in `removeInsignificantPlayers`
- `insert` must be used in `resetPlayerCounts`
- `sort` must not be used in `formatPlayers`
- `updateGoalCountForPlayer` and `List.foldl` must be used in `aggregateScorers`
- `Maybe.map` and `Maybe.withDefault` should be used in `formatPlayer`
- `merge` must be used in `combineGames`

jiegillet replied:
Yeah those are all good ones and they are in the hints (except for not using `sort` in `formatPlayers`, maybe we could add a hint for that).

ceddlyburge replied:
Cool, I'll create an issue on the analyser repo for these, and add a hint for `sort`.

jiegillet commented on exercises/concept/top-scorers/src/TopScorers.elm:
```diff
@@ -0,0 +1,38 @@
+
+aggregateScorers : List PlayerName -> Dict PlayerName Int
+aggregateScorers playerNames =
+    Debug.todo "Use List.foldl and updateGoalCountForPlayer to convert the list in to a Dict"
```
```suggestion
    Debug.todo "Use List.foldl and updateGoalCountForPlayer to convert the list into a Dict"
```

jiegillet commented on exercises/concept/top-scorers/src/TopScorers.elm:
```diff
@@ -0,0 +1,38 @@
+
+updateGoalCountForPlayer : PlayerName -> Dict PlayerName Int -> Dict PlayerName Int
+updateGoalCountForPlayer playerName playerGoalCounts =
+    Debug.todo "implement updateGoalCountForPlayer function, to initialise or increment the goalcount for PlayerName"
```
I really like those short explanations, sometimes they are all you need if you want to be fast.

jiegillet commented on exercises/concept/top-scorers/src/TopScorers.elm:
```diff
@@ -0,0 +1,38 @@
+
+resetPlayerGoalCount : PlayerName -> Dict PlayerName Int -> Dict PlayerName Int
+resetPlayerGoalCount playerName playerGoalCounts =
+    Debug.todo "implement resetPlayerGoalCount function, to remove players who have scored less than the threshold number of goals"
```
```suggestion
    Debug.todo "implement resetPlayerGoalCount function, to reset the score of a player"
```

jiegillet commented on exercises/concept/top-scorers/tests/Tests.elm:
```diff
@@ -0,0 +1,82 @@
+tests =
+    describe "Top Scorers"
+        [ describe
+            "1. Aggregate scorers"
```
All the task-level descriptions [should be a single digit](https://github.com/mpizenberg/elm-test-runner/pull/9/files)
```suggestion
            "1"
```
Yet another reminder that we need better docs :)

jiegillet commented on exercises/concept/top-scorers/tests/Tests.elm:
```diff
@@ -0,0 +1,82 @@
+                            )
+            ]
+        , describe
+            "4. Format the goal count for a single player"
```
We should add a test case when the player is missing

jiegillet commented on config.json:
```diff
@@ -271,6 +276,19 @@
+        ],
+        "prerequisites": [
+          "maybe",
+          "lists"
```
I would also add booleans (since we use filter)

jiegillet reviewed (commented):
Nice concept exercise, it has some really cool tasks! 
I left some comments. Make sure the intro/about documents end up the same.

For the record, here is my solution (without looking at the exemplar)
```elm
module TopScorers exposing (..)

import Dict exposing (Dict)
import TopScorersSupport exposing (PlayerName)


updateGoalCountForPlayer : PlayerName -> Dict PlayerName Int -> Dict PlayerName Int
updateGoalCountForPlayer playerName =
    Dict.update playerName (Maybe.map ((+) 1) >> Maybe.withDefault 1 >> Just)


aggregateScorers : List PlayerName -> Dict PlayerName Int
aggregateScorers =
    List.foldl updateGoalCountForPlayer Dict.empty


removeInsignificantPlayers : Int -> Dict PlayerName Int -> Dict PlayerName Int
removeInsignificantPlayers goalThreshold =
    Dict.filter (\_ score -> score >= goalThreshold)


resetPlayerGoalCount : PlayerName -> Dict PlayerName Int -> Dict PlayerName Int
resetPlayerGoalCount playerName =
    Dict.insert playerName 0


formatPlayer : PlayerName -> Dict PlayerName Int -> String
formatPlayer playerName =
    Dict.get playerName >> Maybe.withDefault 0 >> String.fromInt >> (++) (playerName ++ ": ")


formatPlayers : Dict PlayerName Int -> String
formatPlayers =
    Dict.map (\name score -> name ++ ": " ++ String.fromInt score) >> Dict.values >> String.join ", "


combineGames : Dict PlayerName Int -> Dict PlayerName Int -> Dict PlayerName Int
combineGames game1 game2 =
    Dict.merge Dict.insert (\name a b -> Dict.insert name (a + b)) Dict.insert game1 game2 Dict.empty
```

ceddlyburge commented:
Thanks for the detailed review Jeremie, and for coding up the solution (which is super handy to check that it makes sense and is an appropiate level of difficult and things like that). \

I think I've addressed all the feedback, can you take a look and see what you think.

I like your solution as well, it uses a lot of composition and point free style, which is interesting. We don't have a concept exercise for that, but I feel like it is something we should get to soon, and maybe we could even reuse this exercise, so that students get to practice solving the same thing in two different ways and comparing the two.

I also realise that we haven't taught `|>`, which is ubiquitious in Elm and something of an oversight on our part. We should probably have a think about adding it to an earlier exercise, maybe basics-1 or basics-2

jiegillet commented on concepts/dict/about.md:
````diff
@@ -0,0 +1,91 @@
+```elm
+aliceAndBob = Dict.fromList [ ( "Alice", 1 ), ( "Bob", 1 ) ]
+bobAndCedd = Dict.fromList [ ( "Bob", 1 ), ( "Cedd", 1 ) ]
+empty = Dict.merge
+        -- when only in aliceAndBob
+        (\playerName goalCount mergedGoalCounts -> Dict.insert playerName goalCount mergedGoalCounts)
+        -- when in aliceAndBob and bobAndCedd
+        (\playerName aliceAndBobGoalCount bobAndCeddGoalCount mergedGoalCounts -> Dict.insert playerName (aliceAndBobGoalCount + bobAndCeddGoalCount) mergedGoalCounts)
+        -- when only in bobAndCedd
+        (\playerName goalCount mergedGoalCounts -> Dict.insert playerName goalCount mergedGoalCounts)
+        -- the two dicts to merge
+        aliceAndBob
+        bobAndCedd
+        -- the initial state of the merge
+        Dict.empty
+--> Dict.fromList [ ( "Alice", 1 ), ( "Bob", 2 ), ( "Cedd", 1 ) ]
````
I would propose some changes here because students could copy paste the example to solve the exercise. I also made the variable names shorter for readability (and to examples above for consistency).

```suggestion
firstDictCountsDouble = Dict.merge
        -- when only in aliceAndBob
        (\name count mergedCounts -> Dict.insert name (2 * count) mergedCounts)
        -- when in aliceAndBob and bobAndCedd
        (\name aliceAndBobCount bobAndCeddCount mergedCounts -> Dict.insert name (2 * aliceAndBobCount + bobAndCeddCount) mergedCounts)
        -- when only in bobAndCedd
        (\name count mergedCounts -> Dict.insert name count mergedCounts)
        -- the two dicts to merge
        aliceAndBob
        bobAndCedd
        -- the initial state of the merge
        Dict.empty
--> Dict.fromList [ ( "Alice", 2 ), ( "Bob", 3 ), ( "Cedd", 1 ) ]
```

jiegillet commented on concepts/dict/about.md:
````diff
@@ -0,0 +1,91 @@
+
+```elm
+alice = Dict.fromList [ ( "Alice", 0 ) ]
+keyExists = Dict.get 0 alice--> Just "Alice"
````
```suggestion
keyExists = Dict.get 0 alice --> Just "Alice"
```

jiegillet commented on concepts/dict/about.md:
````diff
@@ -0,0 +1,91 @@
+Items can be updated using `update`.
+
+```elm
+alice = Dict.fromList [ ( "Alice", 0 ) ]
+aliceUpperCase = Dict.update 0 (Maybe.map toUpper) alice --> Dict.fromList [ ( "ALICE", 0 ) ]
````
```suggestion
alice = Dict.fromList [ ( 0, "Alice" ) ]
aliceUpperCase = Dict.update 0 (Maybe.map String.toUpper) alice --> Dict.fromList [ ( 0, "ALICE" ) ]
```

jiegillet reviewed (commented):
Awesome, I left one last round of small suggestions. Sorry for the slow response!

> I like your solution as well, it uses a lot of composition and point free style, which is interesting. We don't have a concept exercise for that, but I feel like it is something we should get to soon, and maybe we could even reuse this exercise, so that students get to practice solving the same thing in two different ways and comparing the two.
> 
> I also realise that we haven't taught |>, which is ubiquitious in Elm and something of an oversight on our part. We should probably have a think about adding it to an earlier exercise, maybe basics-1 or basics-2

I tend to use a little too much of point-free style, it makes my Haskell code unreadable, although in Elm it's not too bad!

I think teaching `|>` is already planned in the "partial application" concept #428.

ceddlyburge commented:
> Yeah those are all good ones and they are in the hints (except for not using sort in formatPlayers, maybe we could add a hint for that).
I've addad a hint for this

mpizenberg commented on concepts/dict/about.md:
```diff
@@ -0,0 +1,91 @@
+
+A [`Dict`][dict] in Elm is an immutable dictionary of zero or more key-value pairs.
+
+Type annotations for dicts are be written as follows
```
```suggestion
Type annotations for dicts are written as follows
```

mpizenberg commented on exercises/concept/top-scorers/.docs/instructions.md:
```diff
@@ -0,0 +1,63 @@
+
+First implement `updateGoalCountForPlayer` to initialise or increment the goal count for a player.
+Then implement the `aggregateScorers` function to count how many goals each player has scored, by using `updateGoalCountForPlayer`.
+This function takes a `List PlayerName`(the names of the players that scored the goals, that can contain duplicate player names), and returns a`Dict PlayerName Int` containing all the players in the list along with how many goals they have scored.
```
```suggestion
This function takes a `List PlayerName` (the names of the players that scored the goals, that can contain duplicate player names), and returns a `Dict PlayerName Int` containing all the players in the list along with how many goals they have scored.
```

mpizenberg commented:
I like the exercise! I haven't read all the code in details but I trust @jiegillet to give a better review about that part than I have the time for right now so I'll let him do the "approval review" to merge this.
Thanks a lot!

jiegillet reviewed (commented):
Awesome :)

I found one more typo!

The last thing left to do is making sure that all the `about.md` and `introduction.md` are the same. Be careful in case we committed changes to different ones in the review process.

jiegillet reviewed (approved):
Great work @ceddlyburge, thank you! 🎉
